README.md
capistrano Cookbook
===================
Creates the required infrastructure to enalbe capistrano deployments. Things like
creating a deploy user and ensuring proper group additions etc are made.

Usage
=====

Add the use of the recipes to the node or include them via another cookbook. The recipes all
draw their configs from a node['capistrano'] and node['apps'] attribs from the node / env.

Data Bag
========

```
knife data bag create users
knife data bag create users deploy
```

```
{
  "id": "deploy",
  "groups": [
    "deploy",
    "sysadmin"
  ],
  "home": "/home/deploy",
  "ssh_keys": [
    "ssh-rsa PUBLIC-KEY-HERE",
    "ssh-rsa ANOTHER-PUBLIC-KEY-HERE..."
  ]
}

```

Node Attributes
===============

Example: (node attributes)

```
"apps": {
    "app_name_env": {
        "databases": {
            "mysql": [
                {
                    "name": "db_name",
                    "user": "db_user_env",
                    "pass": "db_pass"
                }
            ]
        },
        "apache": {
            "vhost": {
                "cookbook": "site_database",
                "template": "vhost.conf.erb",
                "docroot": "public/web",
                "aliases": [ "app.com", "www.app.com", "192.237.181.31"],
                "is_canonical": false
            }
        },
    }
}
```

To use the values, ensure the recipes that drive the servies are in the
nodes run list somewhere, either directly on the node, or included via
a site-cookbook recipe.

Example: node.json attribs

```
"run_list": [
    "recipe[capistrano-apps::app_directory]",
    "recipe[capistrano-apps::app_mysql]"
]
```

Example: site cookbook recipe.rb

```
include_recipe "capistrano-apps::app_apache"
include_recipe "capistrano-apps::app_mysql"
include_recipe "capistrano-apps::app_crons"

# from here, the node attributes can be defined here or on the node
```

Default
-------

Configures the deploy user.

Directory
---------

Ensures the deploy user owns a directory somewhere that acts as the root file system
for apps deployed by that user (e.g. /opt/apps.)

Apache
------

Configures the relationship between deploy user/group and apache user/group. Also
this recipe adds the vhosts as defined in the attributes.

```
"apache": {
    "vhost": {
        "cookbook": "custom_site_cookbook",
        "template": "vhost.conf.erb",
        "docroot": "public/web",
        "aliases": [ "app.com", "www.app.com", "192.237.181.31"],
        "is_canonical": false
    }
},
```

MySQL
-----

Creates the mysql databases defined in the attributes.

```
"databases": {
    "mysql": [
        {
            "name": "db_name",
            "user": "db_user_env",
            "pass": "db_pass"
        }
    ]
},
```

Memcache
--------

Creates memcache pools defined in the node attributes.

```
"memcache_pools": {
    "sessions": {
        "name": "sessions",
        "port": "11211",
        "memory": "16"
    },
    "doctrine": {
        "name": "doctrine",
        "port": "11212",
        "memory": "8"
    }
},
```

Crons
-----

Creates cron jobs defined in the node attributes.

Note: Cron commands are altered to use the 'current' directory as the root.

Todo: Enable a cron to define it's root, so we don't need to depend
on the 'app_directory' recipe.

```
"crons": {
    "smsdelivery": {
        "minute": "*/5",
        "hour": "*",
        "day": "*",
        "month": "*",
        "weekday": "*",
        "action": "create",
        "command": "php public/app/console smsqueuebundle:buffer:deliver"
    },
    "cronbundle": {
        "minute": "*/5",
        "hour": "*",
        "day": "*",
        "month": "*",
        "weekday": "*",
        "action": "create",
        "command": "php bin/phing cron:run"
    }
},
```

Full Example
============

```
    "apps": {
      "appname_dev": {
        "properties": {
          "database.name": "db_dev",
          "database.user": "root",
          "database.pass": "",
          "some.app.var": "cho",
        },
        "apache": {
          "vhost": {
            "docroot": "public/web",
            "aliases": [
              "app.dev",
              "www.app.dev"
            ],
            "is_canonical": false,
            "canonical_domain": "www.app.dev",
            "cookbook": "WRAPPER-COOKBOOK-NAME",
            "template": "file.in.wrapper.cookbook.vhost.conf.erb",
            "canonical_url": "www.app.dev",
            "canonical_protocol": "https"
          }
        },
        "databases": {
          "mysql": [
            {
              "name": "app_dev",
              "user": "root",
              "pass": ""
            }
          ],
          "mysql_service": {
            "server_root_password": ""
          }
        }
      }
    },
    "apache": {
      "server_ssl_listen_ip": "192.168.6.60",
      "listen": [
        "*:80",
        "*:443"
      ]
    }
```


CHANGELOG.md
capistrano CHANGELOG
====================

4.0.4
-----

- Default mailto to '' for all cron jobs

4.0.3
-----

- Add ability to override the deploy user via attribute again
- Remove old reference to deploy user data bag
- Fix setfacl calls on directory creation based on attrib settings

4.0.2
-----

- Fix check for skip\_setfacl to ensure setfacl is run in more cases where it should be

4.0.1
-----

- Use chef group not users\_manage to create deploy user, to avoid deleting 'apache' user from deploy group in chef runs to only add it back in the same chef run

4.0.0
-----

- Major bumps in versions in dependant cookbooks
- Added depends for cookbook yum-mysql-community (per mysql cookbook change)

3.0.0
-----

- BREAKING CHANGE: Switch to use cookbook 'users' rather than 'user'
- BREAKING CHANGE: Switch to use cookbook 'memcached' v3.x from v2.x

2.1.2
-----

- Lock apache2 cookbook at v7. v8 introduces breaking changes related to yum and mysql.

2.1.1
-----

- Fix calls to initial 'letsencrypt' email param
- Be sure to stop apache and restart after cert creation to allow us to use the --standalone flag on port 80/443

2.1.0
-----

- Added ability to use 'letsencrypt' to generate SSL certs automatically

2.0.1
-----

- Fix acl apache/deploy group permissions on app directories (there were cases where the default acl's were
  not set on the deploy dir.)

2.0.0
-----

BREAKING CHANGES:

- Using memcached cookbook ~> 2.0.0, which changes cookbook name for config

1.2.1
-----

- work around for mysql2_chef_gem with chef >= 12.5 @see https://github.com/sinfomicien/mysql2_chef_gem/issues/5

1.2.0
-----

- make deploy user config vagrant / chef solo friendly, so that we can use chef solo to configure
  local deploy test environments easily
- move setfacl to the app loop in apache recipe, so that each app directory gets the setfacl call
  rather than the root directory as a whole. this allows for a vagrant share or mount to be used
  for one dir while a native dir used for the next

1.1.1
-----

- Disable the SSL config from the vhost by default. SSL can be used by overriding the template in a separate cookbook.

1.1.0
-----

- Skip running setfacl in situations its not needed, mainly in vagrant env with nfs drives
- Add ability to manage deploy keys per [node,environment,role] with node attributes
- Added support for chef-solo environments without requiring data_bags for users.

1.0.0
-----
- Reworked to be app driven and more segmented, now each service has an individual recipe.

0.1.0
-----
- Initial release of capistrano cookbook.



recipes/app_apache.rb
include_recipe "capistrano-apps::default"

# reference the deploy user variable from the attributes
deploy_user = node['capistrano']['deploy_user']

# check if our deploy user and apache user are the same
apache_equals_deploy = node["apache"]["user"] === deploy_user

# ensure apache and deploy_user are in each others groups for permissions reasons
unless node["apache"].nil?
    # add deploy_user to apache group (helps when deploy needs to
    # clear cache / files created by apache)
    group node["apache"]["group"] do
        members deploy_user
        append true
    end

    # add apache to deploy_user group (this is the common required
    # action to let apache serve files owned by deploy_user)
    group deploy_user do
        members node["apache"]["user"]
        append true
    end
end

# ensure the default site is not enabled
apache_site "default" do
    enable false
end

# configure the apps
node["apps"].each do |app_name,app|

    if app.has_key? 'apache'

        # check if the app asks us to skip this
        skip_setfacl = (app['apache'].has_key? 'skip_setfacl') && app['apache']['skip_setfacl'] == true

        # only setfacl if apache and deploy user are not the same (we're not required if they are) and
        # also if we've not been asked to skip that step manually
        if !apache_equals_deploy && !skip_setfacl
            # ensure that ACL is setup for apps so apache and the deploy user can both interact with
            # the files created by each other. common use case for this is symfony, it creates cache
            # files with no group write, but with acl, groups can then be set to write on it
            #
            # sudo setfacl -R  -m u:apache:rwX -m u:deploy:rwX /opt/apps/app-name_env
            # sudo setfacl -dR -m u:apache:rwX -m u:deploy:rwX /opt/apps/app-name_env
            #
            # always run this, to ensure proper permissions
            execute "capistrano-directory-setfacl-set-initial" do
                action :run
                command "setfacl -R -m u:#{node["apache"]["user"]}:rwX -m u:#{deploy_user}:rwX #{node['capistrano']['deploy_to_root']}/#{app_name}"
                subscribes :run, "directory[#{node['capistrano']['deploy_to_root']}/#{app_name}]", :immediately
            end

            # always run this, to ensure proper permissions
            execute "capistrano-directory-setfacl-set-default" do
                action :run
                command "setfacl -dR -m u:#{node["apache"]["user"]}:rwX -m u:#{deploy_user}:rwX #{node['capistrano']['deploy_to_root']}/#{app_name}"
                subscribes :run, "directory[#{node['capistrano']['deploy_to_root']}/#{app_name}]", :immediately
            end
        end

        if app["apache"].has_key? 'vhost'

            vhost = app["apache"]["vhost"]
            # document root is in the cap deploy path for the app, current symlink, and into the defined app web root
            docroot_full = "#{node['capistrano']['deploy_to_root']}/#{app_name}/current/#{vhost['docroot']}"

            # get decide if we use https or http
            if vhost.has_key? 'canonical_protocol'
                server_canonical_protocol = vhost['canonical_protocol']
            else
                server_canonical_protocol = 'http'
            end

            # build a domain that we can use as a full url
            if vhost.has_key? 'canonical_domain'
                server_canonical_domain = vhost['canonical_domain']
                server_canonical_url = "#{server_canonical_protocol}://#{vhost['canonical_domain']}"
            else
                server_canonical_domain = vhost['canonical_domain']
                server_canonical_url = false
            end

            # used to setup email for SSL cert validation
            if vhost.has_key? 'letsencrypt_email'
                letsencrypt_email = vhost['letsencrypt_email']
            else
                letsencrypt_email = false
            end

            # create virtual host entry for apache
            web_app app_name do
                if vhost.has_key? 'cookbook' and vhost.has_key? 'template'
                    cookbook vhost['cookbook']
                    template vhost['template']
                else
                    template 'vhost.conf.erb'
                end

                if vhost.has_key? 'port'
                    server_port vhost['port']
                else
                    server_port 80
                end

                docroot docroot_full
                server_name app_name
                server_aliases vhost['aliases']
                server_is_canonical vhost['is_canonical']
                server_canonical_domain server_canonical_domain
                server_canonical_url server_canonical_url
                server_canonical_protocol server_canonical_protocol
                enable true
            end

            # runs only if we have been told to run letsencrypt, and if we have a letsencrypt email
            if vhost.has_key? 'letsencrypt'
                if vhost['letsencrypt']
                    if letsencrypt_email

                        # make sure the directory exists, as this script will probably run before the app is installed with capistrano
                        directory docroot_full do
                            action :create
                            owner 'deploy'
                            group 'deploy'
                            recursive true
                        end

                        ssl_cert_domains = [server_canonical_domain] + vhost['aliases']

                        file "#{node['capistrano']['deploy_to_root']}/#{app_name}/shared/letsencrypt.log" do
                            content "#{node['capistrano']['deploy_to_root']}/letsencrypt/current/letsencrypt-auto certonly --standalone --agree-tos --email #{letsencrypt_email} -w #{docroot_full} -d #{ssl_cert_domains.join(' -d ')}"
                            mode '0755'
                            owner 'deploy'
                            group 'deploy'
                        end

                        deploy 'letsencrypt' do
                            repo 'https://github.com/letsencrypt/letsencrypt'
                            deploy_to "#{node['capistrano']['deploy_to_root']}/letsencrypt"
                            action :deploy
                            purge_before_symlink []
                            create_dirs_before_symlink []
                            symlinks({})
                            symlink_before_migrate({})
                        end

                        # create ssl dir for where we place certs
                        %w{ /etc/httpd/ssl }.each do |dirname|
                            directory dirname do
                                action :create
                                recursive true
                            end
                        end

                        cert_install_dir  = "#{node["capistrano"]["letsencrypt"]["cert_dir"]}/#{server_canonical_domain}"

                        # need to halt the web server first
                        log "free port 80/443 for letsencrypt - stop apache2" do
                            notifies :stop, 'service[apache2]', :immediately
                            not_if { File.exists?("#{cert_install_dir}/fullchain.pem") }
                        end

                        execute 'install_ssl_letsencrypt' do
                            command "#{node['capistrano']['deploy_to_root']}/letsencrypt/current/letsencrypt-auto certonly --standalone --agree-tos --email  #{letsencrypt_email} -w #{docroot_full} -d #{ssl_cert_domains.join(' -d ')}"
                            action :run
                            cwd "/etc/httpd/ssl"
                            not_if { File.exists?("#{cert_install_dir}/fullchain.pem") }
                        end

                        log "ensure start apache2" do
                            notifies :start, 'service[apache2]'
                        end

                        # copy the certs to a place we want them to be
                        link "/etc/httpd/ssl/#{server_canonical_domain}.key" do
                            to "#{cert_install_dir}/privkey.pem"
                        end
                        link "/etc/httpd/ssl/#{server_canonical_domain}.crt" do
                            to "#{cert_install_dir}/cert.pem"
                        end
                        link "/etc/httpd/ssl/#{server_canonical_domain}.chain.crt" do
                            to "#{cert_install_dir}/chain.pem"
                        end

                    else
                        file "#{node['capistrano']['deploy_to_root']}/#{app_name}/shared/letsencrypt.log" do
                            content 'We would NOT have run letsencrypt'
                            mode '0755'
                            owner 'deploy'
                            group 'deploy'
                        end
                    end
                end
            end

            # # note: this SSL install may need to be tweaked, but this is an example of what would need to be done
            # # install ssl certs if listening to 443
            # if node['apache']['listen_ports'].include?("443")
            #     # create ssl dir for where we place certs
            #     %w{ /etc/httpd/ssl/ssl.crt /etc/httpd/ssl/ssl.key }.each do |dirname|
            #         directory dirname do
            #             action :create
            #             recursive true
            #         end
            #     end
            #     # install the key file
            #     cookbook_file "/etc/httpd/ssl/ssl.key/#{app_name}.key" do
            #         backup false
            #         source "ssl/server.key" # this is the value that would be inferred from the path parameter
            #         mode "0644"
            #     end
            #     # install server cert
            #     cookbook_file "/etc/httpd/ssl/ssl.crt/#{app_name}.crt" do
            #         backup false
            #         source "ssl/WWW.DOMAIN.COM.crt" # this is the value that would be inferred from the path parameter
            #         mode "0644"
            #     end
            #     # this file is needed too
            #     cookbook_file "/etc/httpd/ssl/ssl.key/Apache_Plesk_Install.txt" do
            #         source "ssl/Apache_Plesk_Install.txt" # this is the value that would be inferred from the path parameter
            #         mode "0644"
            #         backup false
            #     end
            # end
        end
    end
end
